core/trino-main/src/main/java/io/trino/execution/SqlTask.java

```diff
@@ -371,6 +372,16 @@ else if (taskHolder.getTaskExecution() != null) {
             fullGcTime = taskContext.getFullGcTime();
             dynamicFiltersVersion = taskContext.getDynamicFiltersVersion();
         }
+        else {
+            if (state == FINISHED) {
+                // if task FINISHED successfully but taskHolder is not yet updated with SqlTaskExecution or FinalTaskInfo
```

Comment:
so this is kind of race? Why update of `SqlTaskExecution or FinalTaskInfo` doesn't happen before setting state to `FINISHED`?

Reply:
> so this is kind of race? Why update of SqlTaskExecution or FinalTaskInfo doesn't happen before setting state to FINISHED?

Task may transtion to FINISHED before update happens if task is very quick. In time window between task is started and taskHolderReference is updated.

Reply:
I wonder if order could be reversed, e.g: update object before setting state

Reply:
We do not have control over when task completes. As soon as we start it it may complete immediatelly even if we update taskHolderReference just after starting.
And if you suggest that we first update taskHolderReference and then start task - IDK - but [this comment](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/SqlTask.java#L566-L568) suggests we cannot do that.

Reply:
Yeah. It's tricky and there were bugs around this code in the past. Hope we didn't miss anything here